### DISCUSSION TITLE
Export / visualize

### DISCUSSION THREAD

**Discussion by fkgl** [textX/textX#354] [Q&A] [answered]
Hi,
Please what is the equivalent of these command lines in a python class (main.py)

1. textx visualize file.tx file.ent # my graphviz fonctionne but  this is not the case for this command line.
2. textx generate hello.tx --target dot.
3. textx generate example.hello --grammar hello.tx --target dot

**Comment by igordejanovic** [comment, marked as answer]
If I understand the question correctly you are trying to achieve the same effect from Python code what can be done through `textx generate` command?

`textx visualize` is an old command which has been superseded by `textx generate` as visualization is nothing more but generating output in dot, plantUML or some other graph language.

`textx generate` is implemented [here](https://github.com/textX/textX/blob/master/textx/cli/generate.py#L34). You can see that it calls registered generator function [here](https://github.com/textX/textX/blob/master/textx/cli/generate.py#L118). Generators are registered using [registration and discovery API](http://textx.github.io/textX/latest/registration/).

textX has several built-in generators which you can see with `textx list-generators`. To call them see [this part of the doc](<http://textx.github.io/textX/latest/registration/#calling-a-generator>).

For example, to visualize `.tx` files:

```python
from textx import metamodel_for_file
from textx import generator_for_language_target

# Get the metamodel which knows how to parse the given file.
mm = metamodel_for_file('hello.tx')
# Parse the given file with the provided metamodel
model = mm.model_from_file('hello.tx')
# Get the generator from tx file (textx language) to dot
# Generator function is defined here: https://github.com/textX/textX/blob/master/textx/generators.py#L54
generator = generator_for_language_target('textx', 'dot')
# Call the generator
generator(mm, model, '.', overwrite=True, debug=False)
```

**Reply by fkgl** [reply]
Hi @igordejanovic
Thank you for your reply.
I tried the example above and it works. What about this command in Python code:
_**To convert to png run "dot -Tpng -O hello.dot"**_
Becouse I tried the following lines of code:

generator = generator_for_language_target('textx', 'png')
generator(mm, model, '.', overwrite=True, debug=False)

textx.exceptions.TextXRegistrationError: b'No generators registered for language "textx" and target "png".'

**Reply by igordejanovic** [reply]
`dot` is an external command (not part of textX). It is part of GraphViz. You can lookup online how to install GraphViz and how to call external commands from Python.

**Reply by fkgl** [reply]
Thank you for your reply. 
Now it's OK
